utils/treesitter/grammar.js

```diff
@@ -0,0 +1,507 @@
+      seq('package', $.ident, optional($.library_path), $.api_or_impl, ';'),
+
+    import_directive: ($) =>
+      seq('import', $.ident, optional($.library_path), ';'),
+
+    comment: ($) => token(seq('//', /.*/)),
+
+    ident: ($) => /[A-Za-z_][A-Za-z0-9_]*/,
+
+    bool_literal: ($) => choice('true', 'false'),
+
+    numeric_literal: ($) => {
+      // This is using variables because rules are not allowed in token.immediate and token.
+      // https://github.com/tree-sitter/tree-sitter/issues/449
+      const decimal_integer_literal = choice('0', /[1-9](_?[0-9])*/);
+      const hex_digits = /[0-9A-F](_?[0-9A-F])*/;
+      const binary_integer_literal = /0b[01](_?[01])*/;
+      const hex_integer_literal = seq('0x', token.immediate(hex_digits));
+
+      const decimal_real_number_literal = seq(
+        decimal_integer_literal,
+        token.immediate(/\.[0-9](_?[0-9])*/),
+        optional(
+          seq(
+            token.immediate(/e[+-]?/),
+            token.immediate(decimal_integer_literal)
+          )
+        )
+      );
+
+      const hex_real_number_literal = seq(
+        hex_integer_literal,
+        token.immediate(/\.[0-9A-F]+/),
+        optional(seq(token.immediate(/p[+-]?/), token.immediate(hex_digits)))
+      );
+
+      return token(
+        choice(
+          decimal_integer_literal,
+          binary_integer_literal,
+          hex_integer_literal,
+          decimal_real_number_literal,
+          hex_real_number_literal
+        )
+      );
+    },
+
+    // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2015.md#syntax
+    numeric_type_literal: ($) => /[iuf][1-9][0-9]*/,
+
+    _string_content: ($) => token.immediate(/[^\\"]+/),
+
+    escape_sequence: ($) =>
+      token.immediate(
+        seq(
+          '\\',
+          choice(
+            'n',
+            't',
+            'r',
+            "'",
+            '"',
+            '\\',
+            '0',
+            /x[0-9A-F]{2}/,
+            /u\{[0-9A-F]+\}/
+          )
+        )
+      ),
+
+    // TODO: multiline string
+    string_literal: ($) =>
+      seq(
+        '"',
+        repeat(choice($._string_content, $.escape_sequence)),
+        token.immediate('"')
+      ),
+
+    array_literal: ($) =>
+      seq(
+        '[',
+        field('type', $._expression),
+        ';',
+        optional(field('size', $._expression)),
+        ']'
+      ),
+
+    struct_literal: ($) =>
+      seq('{', comma_sep(seq($.designator, '=', $._expression)), '}'),
+
+    struct_type_literal: ($) =>
+      seq('{', comma_sep(seq($.designator, ':', $._expression)), '}'),
+
+    builtin_type: ($) => choice('Self', 'String', 'bool', 'type'),
+
+    literal: ($) =>
+      choice(
+        $.bool_literal,
+        $.numeric_literal,
+        $.numeric_type_literal,
+        $.string_literal,
+        $.struct_literal,
+        $.struct_type_literal
+      ),
+
+    _binding_lhs: ($) => choice($.ident, '_', seq('addr', 'self')),
```

Comment:
`addr self` and `self` are only allowed in the deduced parameter list, which doesn't use this rule.